com.unity.render-pipelines.high-definition/Editor/RenderPipeline/VolumeComponentWithQualityEditor.cs

```diff
@@ -7,15 +11,238 @@ internal abstract class VolumeComponentWithQualityEditor : VolumeComponentEditor
+                {
+                    return false;
+                }
+
+                if (left.settings.Count != right.settings.Count)
+                {
+                    return false;
+                }
+
+                foreach (var key in left.settings.Keys)
+                {
+                    if (!left.settings[key].Equals(right.settings[key]))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            // TODO: Should be an override of GetHashCode (in core)?
+            int Hash(SerializedDataParameter setting)
+            {
+                int hash = setting.GetHashCode();
+
+                unchecked
+                {
+                    hash = 23 * setting.value.GetHashCode();
+                    hash = 23 * setting.overrideState.GetHashCode();
+                }
+
+                return hash;
+            }
+
+            // Save a setting to the quality blob. 
+            public void Save<T>(SerializedDataParameter setting) where T : struct
+            {
+                QualitySetting s;
+                s.state = setting.overrideState.boolValue;
+                s.value = setting.value.GetInline<T>();
+                
+                int key = Hash(setting);
+
+                if (settings.ContainsKey(key))
+                {
+                    // If the setting exists, overwrite the value
+                    settings[key] = s;
+                }
+                else
+                {
+                    // Otherwise, cache a new setting
+                    settings.Add(key, s);
+                }
+            }
+
+            // Attempts to fill a quality setting with its custom value and override state.
+            public void TryLoad<T>(ref SerializedDataParameter setting) where T : struct
+            {
+                if (settings.TryGetValue(Hash(setting), out QualitySetting s))
+                {
+                    setting.value.SetInline((T)s.value);
+                    setting.overrideState.boolValue = s.state;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Scoped quality setting change checker. 
+        /// </summary>
+        public struct QualityScope : IDisposable
+        {
+            bool m_Disposed;
+            VolumeComponentWithQualityEditor m_QualityComponent;
+            QualitySettingsBlob m_Settings;
+
+            // Cache the quality setting
+            public QualityScope(VolumeComponentWithQualityEditor component)
+            {
+                m_Disposed = false;
+                m_QualityComponent = component;
+                m_Settings = m_QualityComponent.SaveCustomQualitySettingsAsObject();
+                EditorGUI.BeginChangeCheck();
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+            }
+
+            void Dispose(bool disposing)
```

Comment:
You can simplify this:
Dispose is called by the using when exiting it.
In some case we want to dispose of the object either because we know we no longer use it (end of using) or because it is garbage collected (call of the finalizer).
Here you have no finalizer so it should be safe to move all the content of the Dispose(bool) into the Dispose().
As such you cannot encounter the case where you try to dispose an already disposed object so you could also simplify the content under this comment.